relay-general/src/protocol/contexts/browser.rs

```diff
@@ -21,3 +24,233 @@ impl BrowserContext {
         "browser"
     }
 }
+
+impl ContextFromUserAgentInfo for BrowserContext {
+    fn from_client_hints(raw_contexts: &RawUserAgentInfo) -> Option<Self> {
+        let browser = parse_client_hint_browser(raw_contexts.sec_ch_ua?)?;
+        let version = raw_contexts.sec_ch_ua_full_version?.to_owned();
+
+        Some(Self {
+            name: Annotated::new(browser.to_string()),
+            version: Annotated::new(version),
+            ..Default::default()
+        })
+    }
+
+    fn from_user_agent(user_agent: &str) -> Option<Self> {
+        let browser = parse_user_agent(user_agent);
+
+        if !is_known(browser.family.as_str()) {
+            return None;
+        }
+
+        Some(Self {
+            name: Annotated::from(browser.family),
+            version: Annotated::from(get_version(&browser.major, &browser.minor, &browser.patch)),
+            ..BrowserContext::default()
+        })
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum Browser {
+    Chrome,
+    Edge,
+    Safari,
+    Brave,
+    Opera,
+    Firefox,
+    Other(String),
+}
+
+impl Browser {
+    fn as_str(&self) -> &str {
+        match self {
+            Self::Chrome => "chrome",
+            Self::Edge => "edge",
+            Self::Firefox => "firefox",
+            Self::Safari => "safari",
+            Self::Brave => "brave",
+            Self::Opera => "opera",
+            Self::Other(s) => s,
+        }
+    }
+}
+
+impl std::string::ToString for Browser {
+    fn to_string(&self) -> String {
+        self.as_str().to_owned()
+    }
+}
+
+/// the sec-ch-ua field looks something like this:
+/// "Not_A Brand";v="99", "Google Chrome";v="109", "Chromium";v="109"
+///
+/// this function attempts to detect the field with the browser, parse it as a browser enum and
+/// return that. if it fails to parse it as a browser, it uses the "other" variant of browser enum
+///
+/// returns None if no browser field detected
+fn parse_client_hint_browser<S: Into<String>>(s: S) -> Option<Browser> {
+    let s = s.into();
+    let items: Vec<&str> = s.split(',').collect();
+
+    for item in items {
+        let item = item.to_lowercase();
+
+        // if it contains one of these then we can know it isn't a browser field. atm chromium
+        // browsers are the only ones supporting client hints.
+        if item.contains("brand")
```

Comment:
Instead of working with `contains`, I would split by `;` and strip the `""` first. After that, you can do `==` checks.

Reply:
Test comment.